### PR TITLE
system.kafka_consumers table to monitor kafka consumers

### PR DESCRIPTION
**Pull request by ilejn**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
system table to monitor kafka consumers

### Documentation entry for user-facing changes

- [x ] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by ilejn** [comment]
Implements https://github.com/ClickHouse/ClickHouse/issues/5128#issuecomment-1516076817

**Comment by ilejn** [comment]
Hello Kseniia, thank you for assigning this PR.
[Just to let you know] that it is not ready for review yet. 
Main gap - librdkafka_statistics is not covered, I hope to handle it soon.

File: tests/integration/test_storage_kafka/test.py [hunk 1]
```diff
@@ -4504,6 +4505,221 @@ def test_block_based_formats_2(kafka_cluster):
...
+                     format_csv_delimiter = '|';
+        """
+    )
+
+    result = instance.query("SELECT * FROM test.kafka ORDER BY a;")
+
+    result_system_kafka_consumers = instance.query(
+        """
+        create or replace function stable_timestamp as
+          (d)->multiIf(d==toDateTime('1970-01-01 00:00:00'), 'never', abs(dateDiff('second', d, now())) < 20, 'now', toString(d));
```
[the remark this answers is not in the document]
Comment (ilejn) [reply tests/integration/test_storage_kafka/test.py:4544]: Actually comparing to now() seems more load-agnostic than comparing to a time that had obtained before.
20 seconds should be enough, but ok, I'll change it to 30.

File: src/Storages/Kafka/StorageKafka.cpp [hunk 2]
```diff
@@ -641,6 +645,19 @@ void StorageKafka::updateConfiguration(cppkafka::Configuration & kafka_config)
         LOG_IMPL(log, client_logs_level, poco_level, "[rdk:{}] {}", facility, message);
     });
 
+    if (!config.has(config_prefix + "." + "statistics_interval_ms"))
+    {
+        kafka_config.set("statistics.interval.ms", "600"); // every 600 milliseconds
```
[the remark this answers is not in the document]
Comment (ilejn) [reply src/Storages/Kafka/StorageKafka.cpp]: I've tweaked 0025-timers.c test https://github.com/ilejn/librdkafka/blob/prof_stat_cb/tests/0025-timers.c
No visible delays even with 1 millisecond interval.
Although
- json is abridged for several first calls (IOW for 2-4 milliseconds)
- I do not have consumers, so the test is not really meaningful (creating many topics did not change anything).

**Comment by ilejn** [comment]
I believe that test failures are not caused by my changes.

File: src/Storages/Kafka/KafkaConsumer.cpp [hunk 3]
```diff
@@ -527,4 +535,60 @@ void KafkaConsumer::storeLastReadMessageOffset()
     }
 }
 
+void KafkaConsumer::setExceptionInfo(const String & text)
```
Comment (filimonov) [inline src/Storages/Kafka/KafkaConsumer.cpp:538]: shouldn't it be used?

**Review by kssenii — APPROVED** [review]

File: src/Storages/Kafka/StorageKafka.h [hunk 4]
```diff
@@ -142,6 +156,8 @@ class StorageKafka final : public IStorage, WithContext
 
     bool streamToViews();
     bool checkDependencies(const StorageID & table_id);
+
+    std::shared_ptr<const String> rdkafka_stat;
```
Comment (filimonov) [inline src/Storages/Kafka/StorageKafka.h:160]: shouldn't that be per consumer? (instead of global per table).

Comment (ilejn) [reply src/Storages/Kafka/StorageKafka.h:160]: The statistic is global (it is complete for every consumer), I do not see why it is better to have individual copy of the same data in all consumers.
@filimonov  Do you have concerns behind this suggestion?

**Review by filimonov — APPROVED** [review]
LGFM.

One case which is not well covered here is when the exception in the materialized view. At that stage there is no good way to attribute the problem with a particular consumer / topic / offset so it's hard to push that information to kafka_consumers table (well may be push that issue to all the consumers iteratively, but that is also non-perfect). 

Maybe we can leave it like that for now. Materialized view errors should be visible in other places (system.errors, system.query_view_log etc) so it's less important

**Comment by ilejn** [comment]
Hello @kssenii ,
master is currently not green, performance comparison failures (mostly aarch64 though) are common.
Should I try to rebase nevertheless?

**Comment by Yerkon** [comment]
Hello @ilejn,
Currently updated clickhouse to ClickHouse (ver. 23.8.2.7). But the table `system.kafka_consumers` currently is empty, despite to kafka engine consumers are working. 
In `system.events` we have kafka statistics

**Comment by Yerkon** [comment]
@ilejn ,
>  Provide additional information about your scenario. Ideally complete reproduction
1.
```

create table kafkaTable
(
    json String
)
    engine = Kafka SETTINGS kafka_broker_list ='some brokers', kafka_topic_list = 'SomeTopic', kafka_group_name = 'Group1', kafka_format = 'JSONAsString', kafka_skip_broken_messages = 1048576, kafka_num_consumers = 1, kafka_max_block_size = 65536, input_format_import_nested_json = 1, kafka_commit_every_batch = 1;
```

2. Also have **MVs** which insert messages from `kafkaTable` to `SomeReplacingMergeTree` table. Messages are procedeed.
3. From `system.metrics`, we have KafkaConsumers = 18
4. From `system.events`, we have some number of real time kafka messages read, rows like `KafkaMessagesRead,KafkaRowsRead, KafkaMessagesPolled`

**Comment by Slach** [comment]
@Yerkon 
could you share
`ls -la /var/lib/clickhouse/metadata/system/`
`grep -C 10 -i kafka -r /var/lib/clickhouse/preprocessed_configs`

**Comment by Yerkon** [comment]
Hello @Slach, problem was already solved. It was some missing grants

File: src/Storages/Kafka/KafkaConsumer.cpp [hunk 5]
```diff
@@ -527,4 +542,64 @@ void KafkaConsumer::storeLastReadMessageOffset()
...
+
+void KafkaConsumer::setExceptionInfo(const String & text)
+{
+    std::lock_guard<std::mutex> lock(exception_mutex);
+    exceptions_buffer.push_back({text, static_cast<UInt64>(Poco::Timestamp().epochTime())});
+}
+
+/*
+ * Needed until
+ * https://github.com/mfontanini/cppkafka/pull/309
```
Comment (azat) [inline src/Storages/Kafka/KafkaConsumer.cpp:558]: Looks like this got merged https://github.com/mfontanini/cppkafka/pull/309
And it worth to update submodule and remove this function.

**Comment by alexey-milovidov** [comment]
@ilejn, you can see that the Kafka library has data races and memory leaks.
Now this feature is in question...

**Comment by ilejn** [comment]
> @ilejn, you can see that the Kafka library has data races and memory leaks. Now this feature is in question...

Hello @alexey-milovidov , librdkafka is not the best library in the world indeed, but I am not sure that I got your point.
Are we speaking about librdkafka in general or about system.kafka_consumers?
Do we have a plan to replace librdkafka?

system.kafka_consumers supposed to help customers to find out and mitigate [librd]kafka issues. If librdkafka does have memory leaks, extending system.kafka_consumers to cover these types of issues may be a good starting point to fixing them.
If we are aware of issues with system.kafka_consumers , I willingly fix them.

**Comment by filimonov** [comment]
> that the Kafka library has data races and memory leaks.

Your criticism regarding librdkafka might not be unfounded, yet it's essential to recognize that Kafka's architecture adheres to the principle of "simple broker, complex consumer." This design philosophy necessitates a multitude of sophisticated features that need to be supported on the client side. The complexity inherent in these features, especially given their multi-threaded implementation in pure C, and attempts to overlay a simple external API atop them, naturally give rise to various issues. While librdkafka may not be flawless, it is the official client, supporting all (or nearly all) Kafka features.

As of my knowledge, there are no significantly better alternatives now in c++ world. The temptation to "write your own kafka client with blackjack etc" does exist, yet this usually dissipates upon delving into the details and grasping the scale of the challenge at hand.

Creating something from scratch, like a client for the Kafka 0.8 protocol without complex authentication methods, is feasible within a reasonable timeframe. However, developing a universal client that supports Kafka protocol 1.0 features and beyond is a task that spans months or years, requiring ongoing maintenance thereafter.

To give you an idea of the scale:

- The librdkafka library itself comprises about 190K lines of code (LOC).
- The Go client, [sarama by IBM](https://github.com/IBM/sarama), has around 50K LOC.
- The [kafka-python](https://github.com/dpkp/kafka-python) library is approximately 25K LOC.
- [tulios/kafkajs](https://github.com/tulios/kafkajs), another JavaScript client, totals about 50K LOC.
- [rskafka by influxdata](https://github.com/influxdata/rskafka), described as 'A minimal Rust client for Apache Kafka', is around 15K LOC.
- The Rust client for Apache Kafka, [kafka-rust](https://github.com/kafka-rust/kafka-rust/), has about 10K LOC.

(LOC stats from https://ghloc.vercel.app/)

The client implementations with fewer lines of code usually lack support for some important or modern features of Apache Kafka.

File: src/Storages/Kafka/KafkaConsumer.cpp (hunk 5 above)
Comment (ilejn) [reply src/Storages/Kafka/KafkaConsumer.cpp:558]: In process.

**Pull request merged**